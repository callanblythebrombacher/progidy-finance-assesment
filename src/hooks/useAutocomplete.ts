import {Dispatch, SetStateAction, useEffect, useState} from 'react';
import {AutoCompleteData} from '../interfaces/atom.interfaces';
import {Filter} from '../utils/filter';
import {Normalize} from '../utils/normalize';
import {NormalizedCountryArrayItem} from '../interfaces/normalize.interface';

// Initialize instances of Filter and Normalize classes
const filter = new Filter();
const normalize = new Normalize();

/**
 * Custom hook for handling autocomplete functionality.
 * @param setSearchQuery - Setter function for the search query state.
 * @param searchQuery - Current search query string.
 * @param searchData - Data to be used for autocomplete suggestions.
 * @param pickerValue - Value indicating the picker selection.
 * @param countryData - Array of normalized country data.
 * @param setRowData - Setter function for row data.
 * @param setPage - Setter function for the current page.
 * @returns Object containing autoCompleteData, setAutoCompleteData, and handleItemSelection function.
 */
const useAutocomplete = (
  setSearchQuery: Dispatch<SetStateAction<string>>,
  searchQuery: string,
  searchData: AutoCompleteData,
  pickerValue: number,
  countryData: NormalizedCountryArrayItem[] | null,
  setRowData: (data: any) => void,
  setPage: Dispatch<SetStateAction<number>>,
) => {
  // State to manage autocomplete data
  const [autoCompleteData, setAutoCompleteData] = useState<
    AutoCompleteData | undefined
  >([]);

  // Function to filter searchData based on query
  const handleFilterSearch = (
    query: string,
    searchData: AutoCompleteData,
  ): AutoCompleteData => {
    const trimmedQuery = query.trim().toLowerCase();
    return searchData.filter(item =>
      item.listTitle.toLowerCase().includes(trimmedQuery),
    );
  };

  // Function to handle item selection from autocomplete suggestions
  const handleItemSelection = (itemTitle: string) => {
    setSearchQuery(itemTitle); // Update search query with selected item
    setAutoCompleteData([]); // Clear autocomplete suggestions
  };

  useEffect(() => {
    // Ensure searchData is available and not null/undefined
    if (!searchData) return;

    // Filter searchData based on searchQuery
    const filteredData = handleFilterSearch(searchQuery, searchData);

    // Handle a selection event and update table data
    if (
      filteredData.length === 1 &&
      searchQuery === filteredData[0].listTitle
    ) {
      handleItemSelection(filteredData[0].listTitle); // Clear autocomplete data on item selection
      const filteredCountryData = filter.getFilteredCountryData(
        pickerValue,
        searchQuery,
        countryData,
      );
      const rowConfig = normalize.getRowConfig(filteredCountryData);
      setRowData(rowConfig); // Set row data based on filtered configuration
    }

    // Update autoCompleteData based on filtered results
    if (
      filteredData.length > 0 &&
      searchQuery !== '' &&
      searchQuery !== filteredData[0].listTitle
    ) {
      setAutoCompleteData(filteredData);
      setPage(0); // Reset page to 0 when search query changes
    } else {
      setAutoCompleteData(undefined);
    }
  }, [searchQuery, searchData]); // Dependencies for useEffect

  // Return autoCompleteData, setAutoCompleteData, and handleItemSelection function
  return {
    autoCompleteData,
    setAutoCompleteData,
    handleItemSelection,
  };
};

export default useAutocomplete;
