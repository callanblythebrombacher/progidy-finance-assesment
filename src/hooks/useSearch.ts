import {useState} from 'react';

const useSearch = (initialQuery: string = '') => {
  const [searchQuery, setSearchQuery] = useState(initialQuery);

  const handleSearch = (text: string) => setSearchQuery(text);

  const handleClearInput = () => setSearchQuery('');

  return {
    setSearchQuery,
    searchQuery,
    handleSearch,
    handleClearInput,
  };
};

export default useSearch;
