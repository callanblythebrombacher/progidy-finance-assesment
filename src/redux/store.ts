import {configureStore} from '@reduxjs/toolkit';
import countryReducer from './reducers/country.slice.ts';

export const store = configureStore({
  reducer: {
    countryReducer,
  },
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;
