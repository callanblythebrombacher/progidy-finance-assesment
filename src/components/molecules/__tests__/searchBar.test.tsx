import React from 'react';
import {shallow} from 'enzyme';
import {TextInput, View} from 'react-native';
import {IconButton} from 'react-native-paper';
import {useSelector} from 'react-redux';
import {SearchBar} from '../searchBar.molecure.tsx';
import {Autocomplete} from '../../atoms/autocomplete.atom';
import useSearch from '../../../hooks/useSearch';
import useAutocomplete from '../../../hooks/useAutocomplete';
import {RootState} from '../../../redux/store';

import {jest, describe, beforeEach, afterEach, it, expect} from '@jest/globals';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

jest.mock('../src/hooks/useSearch');
jest.mock('../src/hooks/useAutocomplete');
jest.mock('../src/atoms/autocomplete.atom', () => ({
  Autocomplete: () => 'Autocomplete',
}));

describe('SearchBar Component', () => {
  const mockUseSelector = useSelector as jest.Mock;
  const mockUseSearch = useSearch as jest.Mock;
  const mockUseAutocomplete = useAutocomplete as jest.Mock;

  const mockSearchData = [{id: 1, name: 'Test Country'}];
  const mockClearFilteredSearch = jest.fn();
  const mockSetRowData = jest.fn();
  const mockSetPage = jest.fn();

  beforeEach(() => {
    mockUseSelector.mockReturnValue(mockSearchData);
    mockUseSearch.mockReturnValue({
      searchQuery: '',
      handleSearch: jest.fn(),
      handleClearInput: jest.fn(),
      setSearchQuery: jest.fn(),
    });
    mockUseAutocomplete.mockReturnValue({
      autoCompleteData: [],
      handleItemSelection: jest.fn(),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with given props', () => {
    const wrapper = shallow(
      <SearchBar
        searchData={mockSearchData}
        clearFilteredSearch={mockClearFilteredSearch}
        searchQuery=""
        pickerValue=""
        setRowData={mockSetRowData}
        setPage={mockSetPage}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders TextInput and IconButton components', () => {
    const wrapper = shallow(
      <SearchBar
        searchData={mockSearchData}
        clearFilteredSearch={mockClearFilteredSearch}
        searchQuery=""
        pickerValue=""
        setRowData={mockSetRowData}
        setPage={mockSetPage}
      />,
    );

    expect(wrapper.find(TextInput)).toHaveLength(1);
    expect(wrapper.find(IconButton)).toHaveLength(0); // No icon button when searchQuery is empty

    wrapper.setProps({searchQuery: 'Test'});
    expect(wrapper.find(IconButton)).toHaveLength(1); // Icon button when searchQuery is not empty
  });

  it('calls handleSearch on TextInput change', () => {
    const mockHandleSearch = jest.fn();
    mockUseSearch.mockReturnValue({
      searchQuery: '',
      handleSearch: mockHandleSearch,
      handleClearInput: jest.fn(),
      setSearchQuery: jest.fn(),
    });

    const wrapper = shallow(
      <SearchBar
        searchData={mockSearchData}
        clearFilteredSearch={mockClearFilteredSearch}
        searchQuery=""
        pickerValue=""
        setRowData={mockSetRowData}
        setPage={mockSetPage}
      />,
    );

    wrapper.find(TextInput).simulate('changeText', 'Test');
    expect(mockHandleSearch).toHaveBeenCalledWith('Test');
  });

  it('calls handleClearInput and clearFilteredSearch on IconButton press', () => {
    const mockHandleClearInput = jest.fn();
    mockUseSearch.mockReturnValue({
      searchQuery: 'Test',
      handleSearch: jest.fn(),
      handleClearInput: mockHandleClearInput,
      setSearchQuery: jest.fn(),
    });

    const wrapper = shallow(
      <SearchBar
        searchData={mockSearchData}
        clearFilteredSearch={mockClearFilteredSearch}
        searchQuery="Test"
        pickerValue=""
        setRowData={mockSetRowData}
        setPage={mockSetPage}
      />,
    );

    wrapper.find(IconButton).simulate('press');
    expect(mockHandleClearInput).toHaveBeenCalled();
    expect(mockClearFilteredSearch).toHaveBeenCalled();
  });

  it('renders Autocomplete component with correct props', () => {
    const mockAutoCompleteData = [{id: 1, name: 'Test Suggestion'}];
    const mockHandleItemSelection = jest.fn();
    mockUseAutocomplete.mockReturnValue({
      autoCompleteData: mockAutoCompleteData,
      handleItemSelection: mockHandleItemSelection,
    });

    const wrapper = shallow(
      <SearchBar
        searchData={mockSearchData}
        clearFilteredSearch={mockClearFilteredSearch}
        searchQuery="Test"
        pickerValue=""
        setRowData={mockSetRowData}
        setPage={mockSetPage}
      />,
    );

    const autocomplete = wrapper.find(Autocomplete);
    expect(autocomplete).toHaveLength(1);
    expect(autocomplete.prop('autocompleteData')).toEqual(mockAutoCompleteData);
    expect(autocomplete.prop('onItemPress')).toEqual(mockHandleItemSelection);
  });
});
