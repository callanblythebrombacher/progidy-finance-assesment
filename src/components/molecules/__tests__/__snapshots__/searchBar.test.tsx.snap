// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component renders correctly with given props 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "width": 210,
      "zIndex": 4,
    }
  }
>
  <TextInput
    onChangeText={[MockFunction]}
    placeholder="Search"
    style={
      {
        "backgroundColor": "white",
        "borderColor": "#000000",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 50,
        "padding": 10,
      }
    }
    value=""
  />
  Autocomplete
</View>
`;
